File: src/main/java/tech/bobliu/assignment08/service/ThreadServiceImpl.java
package tech.bobliu.assignment08.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import tech.bobliu.assignment08.model.Post;
import tech.bobliu.assignment08.model.Thread;
import tech.bobliu.assignment08.model.User;
import tech.bobliu.assignment08.repository.PostRepository;
import tech.bobliu.assignment08.repository.ThreadRepository;

import java.util.List;

@Service
public class ThreadServiceImpl implements ThreadService {

    private final ThreadRepository threadRepository;
    private final PostRepository postRepository;

    @Autowired
    public ThreadServiceImpl(ThreadRepository threadRepository, PostRepository postRepository) {
        this.threadRepository = threadRepository;
        this.postRepository = postRepository;
    }

    @Override
    public List<Thread> getAllThreads() {
        return threadRepository.findAll();
    }

    @Override
    public Thread getThread(Long id) {
        return threadRepository.findById(id).orElse(null);
    }

    @Override
    public Thread createThread(String title, String content, User sender) {
        Thread thread = new Thread();
        thread.setTitle(title);
        thread.setSender(sender);

        Post post = new Post();
        post.setSender(sender);
        post.setContent(content);
        post.setThread(thread);

        thread.getPosts().add(post);
        return threadRepository.save(thread);
    }

    @Override
    public Post reply(Long threadId, String content, User sender, Long replyToId) {
        Thread thread = threadRepository.findById(threadId)
                .orElseThrow(() -> new IllegalArgumentException("帖子不存在"));

        Post parentPost = null;
        if (replyToId != null) {
            parentPost = postRepository.findById(replyToId).orElse(null);
        }

        Post post = new Post();
        post.setSender(sender);
        post.setContent(content);
        post.setThread(thread);
        post.setParent(parentPost);

        return postRepository.save(post);
    }
}

File: src/main/kotlin/tech/bobliu/assignment08/service/ThreadServiceImpl.kt
package tech.bobliu.assignment08.service

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import tech.bobliu.assignment08.model.Post
import tech.bobliu.assignment08.model.Thread
import tech.bobliu.assignment08.model.User
import tech.bobliu.assignment08.repository.PostRepository
import tech.bobliu.assignment08.repository.ThreadRepository

@Service
class ThreadServiceImpl @Autowired constructor(
    private val threadRepository: ThreadRepository,
    private val postRepository: PostRepository
) : ThreadService {

    override fun getAllThreads(): List<Thread> {
        return threadRepository.findAll()
    }

    override fun getThread(id: Long): Thread? {
        return threadRepository.findById(id).orElse(null)
    }

    override fun createThread(title: String, content: String, sender: User): Thread {
        val thread = Thread(title = title, sender = sender)
        val post = Post(sender = sender, content = content, thread = thread)
        thread.posts.add(post)
        return threadRepository.save(thread)
    }

    override fun reply(threadId: Long, content: String, sender: User, replyToId: Long?): Post {
        val thread = threadRepository.findById(threadId).orElseThrow { IllegalArgumentException("帖子不存在") }
        var parentPost: Post? = null
        if (replyToId != null) {
            parentPost = postRepository.findById(replyToId).orElse(null)
        }
        val post = Post(sender = sender, content = content, thread = thread, parent = parentPost)
        return postRepository.save(post)
    }
}
